### ISSUE TITLE
Application unresponsive after some time.

### ISSUE THREAD

**Issue by pennyloafers** [aseprite/aseprite#4544]
Describe your bug report or feature request here
Aseprite becomes unresponsive and a CPU core goes 100% usage after leaving application open for long period.
What I mean by a long time is: maybe a day or two, computer is also suspended a few times with other use. There is no specific interaction with the application just that the laptop starts to sweat and I notice that its aseprite eating the CPU and unresponsive.

By its behavior to the system seems like the application is busy spinning in a loop.

### Aseprite and System version

* Aseprite version: 1.3.7
* System: Ubuntu 22.04.4,

**Comment by Liebranca** [comment]
I briefly looked into this and gathered a few things, but not enough for producing a patch. I have some questions:

- Does the application stay unresponsive indefinitely, or does it go back to normal after a while?
- Is memory usage also abnormal, and if so, is swapping enabled?
- Does this happen on the default desktop environment/window manager setup?
 
<br>

Thank you.
#
As for what I've gathered:

For context, if there are no interactions, then [this call](https://github.com/aseprite/aseprite/blob/4df11ac1e31760b7236c2c627a2fff44961e1684/src/ui/manager.cpp#L434) will block through `XNextEvent` and wait for input; so unless something's wrong with X11 itself, then idling for however long should be safe in theory.

But it occurred to me that it may be possible to overwhelm the event queue if the process stalls for a few seconds when being woken up, as can happen while the system loads memory it swapped out during idle time.

What I tried is simulating such a stall by placing two conditional breakpoints from gdb:

- One [here](https://github.com/aseprite/laf/blob/a971ccdddfaf7c7330fe8c3c2fef9cb449c99796/os/x11/event_queue.cpp#L103), checking for `events > 64`, with a command attached for echoing out the number of events and automatically resuming execution.
- The other [here](https://github.com/aseprite/laf/blob/a971ccdddfaf7c7330fe8c3c2fef9cb449c99796/os/x11/event_queue.cpp#L137), checking for `m_events.m_queue.empty()` and `events == 1` for the condition, and no command attached.

When the latter was reached, I manually sent some quick mouse motion/mouse button events to the main window.

Uppon resuming execution, about as expected, anywhere between a couple hundred to a thousand and a few events were received all at once, causing a second (this time unintended) momentary freeze on my end. The more extreme version of this experiment can be consistently achieved by repeatedly resizing and moving the window during the stall.

But conditional breakpoints can be slow, on top of the performance penalty imposed by the debugger, and this has to be accounted for. I set the second point to also automatically resume execution, and that was fairly sluggish, but there was no outright freeze.

In conclusion: Simulating a stall directly in the code, paired with a proper benchmark so as to eliminate guesswork, would give a much clearer picture. I am not certain whether there is some correlation between the behaviour I observed and this particular issue. Perhaps I barked up the wrong tree, but if not, it's something to keep an eye on.

**Comment by pennyloafers** [comment]
One core is at 100%, memory is stable. It never recovers. 

![image](https://github.com/user-attachments/assets/2cc84718-9fb3-4803-8b77-46a5dbbed5ef)

If I resize the window returning to the original size shows black.
![image](https://github.com/user-attachments/assets/d055cfae-ab95-4509-b704-01e37893de59)

It took about 20 minutes, I just opened the app with the inner preview window and an animation preview playing. when I touched the mouse to wake the screen I saw the animation preview play the loop then stop, becoming unresponsive.

I'm looking to get gdb running myself for the practice.

**Comment by Liebranca** [comment]
My bad: I assumed from the initial description that the application was completely idle, and thus discarded that the problem could be elsewhere. Was the animation preview also running during the previous occurrences?

**Comment by pennyloafers** [comment]
Sorry for the delay

> My bad: I assumed from the initial description that the application was completely idle, and thus discarded that the problem could be elsewhere. Was the animation preview also running during the previous occurrences?

No worries, the second occurrence I thought I did not have any animation playing (i think...), but for all occurrences i did. This is just a little easier to check it to see if it crashed..

**Comment by pennyloafers** [comment]
It happened again, and I was walking around with gdb. The app was frozen and when i halted the program. It typically stopped deep under  `void PlayState::onPlaybackTick()`. So I decided to stop the playback state by modifying the check
```
gdb) set var m_playback.m_playMode = 4
(gdb) print m_playback.m_playMode
$3 = doc::Playback::Stopped
```
I then tried to continue to find the real issue, but this seemed to fix the deadlock.  I assume `m_editor->stop();` could have done a lot to the system state. So I will have to wait again to give it another shot.

**Comment by Liebranca** [comment]
Hi, thank you very much for the report!

Knowing on which functions the freezing happens already helps a lot in narrowing down where the bug might be. If it happens again, you can get a backtrace (`bt` command with gdb) to see a list of functions invoked up to that point, that could narrow it down even further.

But anyway, going by your description, and looking at `PlayState::onPlaybackTick()`, it occurs to me that since none of the values involved in the calculation of `m_nextFrameTime` are checked against a limit, the following loop could stall if this difference is larger than expected.

I find this somewhat unlikely, but it's a possibility, so it's best to be sure. I'll make a note to test if offsetting the timestamps by a few days of uptime has any noticeable effect.
